### DISCUSSION TITLE
约定式路由的额外信息如何定义？

### DISCUSSION THREAD

**Discussion by tolerance-go** [umijs/umi#9263] [Q&A]
我在使用 umi max + access，layout，initialState 插件，发现一些问题：
1. 约定式路由的额外信息如何定义？指 access 等
2. config.routes 的异步返回方式

为什么会使用：
我在项目中需要对用户的路由权限进行设计

思考：
路由权限设计，菜单栏显示，路由数据分开来看待：
1. 导航菜单项是否可见
4. 路由是否被注册
5. 组件是否可见
  a. 路由页面是否可见
  b. 页面内组件是否可见

方案：
通过 umi runtime-api，我们处理 clientRoutes data，交给 layout 和 access 插件处理

umi 阶段
我根据 patchClientRoutes 过滤，可以实现 【2】，同时挂载 access-meta 信息，可以间接实现 【3-a】

插件阶段
access 插件
接收 initialState 定义系统 access 能力列表，通过暴露出去的 API 可以实现【3-b】
layout 插件
接收 initialState 可以异步过滤菜单后渲染，实现【1】

问题：
1. 除了 patchClientRoutes 不知道有没有其他更好的方案，总觉写起来怪怪的

```tsx
const routerMeta: Record<string, LayoutPluginExtendClientRoute> = {
  [PATHS.DASHBOARD]: {
    // access: 'canSeeAdmin',
  },
  '/discuss': {},
  '/versions': {},
  '/teams': {},
  '/widgets': {},
  '/widgets/install': {},
  '/widgets/publish': {},
};

export const patchClientRoutes: RuntimeConfig['patchClientRoutes'] = ({
  routes,
}: {
  routes: (IClientRoute & LayoutPluginExtendClientRoute)[];
}) => {
  // [pro-layout:[global-layout:[...routes]], openapi]
  routes[0]?.routes?.[0]?.routes?.forEach((route) => {
    Object.assign(route, routerMeta[`/${route.path}`]);
  });
};
```

2. 并且和配置式一样都不支持异步，如果有这么方面需求该怎么做呢？

其他：
如果可以封装一个异步处理路由数据的插件，同时接收 initialState，如果不需要异步，插件还可以约定一下路由文件导出额外信息

```tsx
// app.ts
const getClientRoutes = async (routes, initialState) => {
    // patchClientRoutes
    return routes
}

// page-a.tsx
export default () => {}
export const defineRouteData({
    access: '..',
})
```

这个方案不知道可用性高不高-。-

**Comment by fz6m** [comment]
试试 https://umijs.org/docs/api/api#useoutletcontext ？

**Reply by tolerance-go** [reply]
谢谢回复，可以给一个例子吗

不过我在官网找到一个例子可以异步处理 routes

```tsx
let extraRoutes;
 
export function patchClientRoutes({ routes }) {
  // 根据 extraRoutes 对 routes 做一些修改
  patch(routes, extraRoutes);
}
 
export function render(oldRender) {
  fetch('/api/routes')
    .then((res) => res.json())
    .then((res) => {
      extraRoutes = res.routes;
      oldRender();
    });
}
```
